### PR TITLE
Don't extend collection types in ArrayOps

### PR DESCRIPTION
**Pull request by szeiger**
+ more ArrayOps methods and optimizations

See commit comments for details.

### CODE REVIEW

**Comment by lrytz** [comment]
Nice! What are your thoughts for keeping the "interface" (available methods) in synch? https://github.com/scala/collection-strawman/issues/465

**Comment by julienrf** [comment]
This PR makes type inference fail on arrays usage. Also, what do you think of implementing the optimized versions of reduction operations on `WrappedArray` instead (which inherits from `IterableOps`)?

**Comment by szeiger** [comment]
There's one type inference failure in the tests. I have not been able to figure out why this change makes it fail. The example we have is something like `Array("") ++ Array("", null)`. ArrayOps should have exactly the same methods as before (i.e. a single `++` method that takes a `ClassTag`; the old implementation had another without the `ClassTag` which was defined in a superclass and therefore shadowed) but the failure is new. Note that the general problem already existed before this change, we just didn't have any test for it: `Iterable("") ++ Array("", null)` fails in the same way even though it does not involve `ArrayOps`.

**Comment by szeiger** [comment]
We could additionally make `WrappedArray` use more optimized implementations but the primary place for them should be `ArrayOps` because it is a pure value class. Because of the specialization `WrappedArray` always needs to allocate an instance (unless you implement a method in all of the specialized subclasses, which was probably the reason for the duplicate `apply` and `update` implementations in the old `ArrayOps`).

**Comment by szeiger** [comment]
Something weird is going on with the `Array.apply` inference problem. The good news is that it works when we're using `Predef` and the `scala` namespace:

```
[info] Running scala.tools.nsc.MainGenericRunner -usejavacp
Welcome to Scala 2.13.0-20180220-202434-37b204e (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_102).
Type in expressions for evaluation. Or try :help.

scala> Array("") ++ Array("", null)
res0: Array[String] = Array("", "", null)

scala> Iterable("") ++ Array("", null)
res1: Iterable[String] = List("", "", null)
```

The problem only occurs when testing in the strawman namespace. My guess would be that some special case handling in the compiler is getting in the way. For example, you can see in `-Ytyper-debug` output that `arrayToWrappedArray` is always considered as an implicit conversion even when it is not in scope and not imported.

**Comment by szeiger** [comment]
Ticket for the overload problem: https://github.com/scala/bug/issues/10746

The reason why it works from Predef is that we still have the separate implicits for converting primitive and reference arrays to `WrappedArray`, whereas in strawman there is only a single conversion method.

File: collections/src/main/scala/strawman/collection/mutable/ArrayBuilder.scala [hunk 1]
```diff
@@ -69,12 +96,13 @@ object ArrayBuilder {
...
-  final class ofRef[T <: AnyRef : ClassTag] extends ArrayBuilder[T] {
+  final class ofRef[T <: AnyRef](implicit ct: ClassTag[T]) extends ArrayBuilder[T] {
 
-    private var elems: Array[T] = _
+    protected var elems: Array[T] = _
 
     private def mkArray(size: Int): Array[T] = {
-      val newelems = new Array[T](size)
+      // Anything short of this will make Dotty create an Array[Object]:
+      val newelems = java.lang.reflect.Array.newInstance(ct.runtimeClass, size).asInstanceOf[Array[T]]
```
Comment (szeiger) [inline collections/src/main/scala/strawman/collection/mutable/ArrayBuilder.scala]: This is a strange one. I was unable to come up with a reproduction in the REPL but it fails consistently here. The `ClassTag` is the one for `Tuple2` but I end up with an `Array[Object]` when I use `new Array` or `Array.ofDim` (even passing the `ClassTag` explicitly).

File: collections/src/main/scala/strawman/collection/ArrayOps.scala [hunk 2]
```diff
@@ -79,24 +536,138 @@ class ArrayOps[A](val xs: Array[A]) extends AnyVal
...
+      b(i) = ((xs(i), i))
+      i += 1
+    }
+    b
+  }
+
+  /** A copy of this array with an element appended. */
+  def appended[B >: A : ClassTag](x: B): Array[B] = {
+    val dest = new Array[B](xs.length + 1)
+    Array.copy(xs, 0, dest, 0, xs.length)
```
Comment (lrytz) [inline collections/src/main/scala/strawman/collection/ArrayOps.scala]: @mkeskells suggests using `Arrays.copyOf(xs, xs.length+1, classTag[B])` (https://github.com/scala/scala/pull/6105#issuecomment-345505209)

Comment (szeiger) [reply collections/src/main/scala/strawman/collection/ArrayOps.scala]: Good idea. We can also optimize all `slice`-based operations with `copyOfRange`.

Comment (lrytz) [reply collections/src/main/scala/strawman/collection/ArrayOps.scala]: Do you think it's possible to optimize for the case where `B` is `A`, could we have an overload that avoids materializing the ClassTag?

File: collections/src/main/scala/strawman/collection/mutable/ArrayBuilder.scala (hunk 1 above)
Comment (allanrenucci) [reply collections/src/main/scala/strawman/collection/mutable/ArrayBuilder.scala]: I opened lampepfl/dotty#4037

**Pull request merged**